flask/flask-launcher/src/main/java/net/corda/flask/launcher/Launcher.java

```diff
@@ -152,18 +154,40 @@ final int launch(Manifest manifest, List<String> jvmArgs, List<String> javaAgent
             }
             builder.getCliArgs().addAll(args);
             beforeChildJvmStart(builder);
-            try(LockFile processLock = LockFile.acquire(cache.getPidFile(), false)) {
-                builder.getProperties().put(Flask.JvmProperties.PID_FILE, cache.getPidFile());
-                Path heartbeatAgentPath = extractedLibraries.get(
-                        manifest.getMainAttributes().getValue(Flask.ManifestAttributes.HEARTBEAT_AGENT_HASH));
-                builder.getJvmArgs().add("-javaagent:" + heartbeatAgentPath);
-                int returnCode = builder.exec();
-                afterChildJvmExit(returnCode);
-                cache.touchLibraries();
-                return returnCode;
-            } finally {
-                Files.delete(cache.getPidFile());
-            }
+            builder.getProperties().put(Flask.JvmProperties.PID_FILE, cache.getPidFile());
+            Path heartbeatAgentPath = extractedLibraries.get(
+                    manifest.getMainAttributes().getValue(Flask.ManifestAttributes.HEARTBEAT_AGENT_HASH));
+            builder.getJvmArgs().add("-javaagent:" + heartbeatAgentPath);
+            LockFile processLock = LockFile.acquire(cache.getPidFile(), false);
+            Process process = builder.build().inheritIO().start();
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                if(process.isAlive()) {
+                    process.destroy();
+                    try {
+                        long timeout = Long.parseLong(
+                                System.getProperty(
+                                        Flask.JvmProperties.KILL_TIMEOUT_MILLIS,
+                                        Flask.Constants.DEFAULT_KILL_TIMEOUT_MILLIS));
+                        process.waitFor(timeout, TimeUnit.MILLISECONDS);
+                    } catch (InterruptedException ie) {
+                        log.error(ie.getMessage(), ie);
+                    } finally {
+                        if (process.isAlive()) {
+                            process.destroyForcibly();
+                        }
+                        processLock.close();
+                    }
+                }
+                try {
+                    Files.delete(cache.getPidFile());
+                } catch (IOException ex) {
+                    log.warn(ex.getMessage(), ex);
+                }
+            }));
+            int returnCode = process.waitFor();
+            afterChildJvmExit(returnCode);
+            cache.touchLibraries();
```

Comment:
Do we want
```java
int returnCode;
try {
    returnCode = process.waitFor();
    afterChildJvmExit(returnCode);
} finally {
    cache.touchLibraries();
}
```
here? Because `waitFor()` can throw `InterruptedException`, which would skip the `cache.touchLibraries()` step.

Does Flask also need an `afterChildJvmError(Exception e)` handler? (We cannot expect to handle `Error` or `Throwable`.)

Reply:
it isn't super-important to execute `cache.touchLibraries();` (the worst thing that can happen is that those libraries will be evicted from the cache and will need to be extracted again the next time the program runs). I can simply move that step at the end of the shutdown hook. 
I don't think we can create an `afterChildJvmError(Exception e)` handler since we cannot access exceptions thrown by the child process (it isn't the case for flask but a child process might not even be a Java process at all)

Reply:
I'm not suggesting handling exceptions thrown _by_ the child process. I am concerned about the case where `process.waitFor()` throws an exception and so skips the `afterChildJvmExit()` hook.

Reply:
ok, but the only exception that can be thrown by `process.waitFor` is `InterruptedException` when the thread is interrupted by another thread, but there is no other thread in Flask that will ever attempt such a thing...